### PR TITLE
Sd clear - settings

### PR DESCRIPTION
**Pull request by jankrzempek**

### CODE REVIEW

File: AirCasting/CreateSessionViews/SelectPeripheralView.swift [hunk 1]
```diff
@@ -95,7 +96,7 @@ struct SelectPeripheralView: View {
     
     var titleLabel: some View {
         var title: Text
-        if syncMode == true {
+        if syncMode == true || SDClearingRouteProcess {
             title = Text(Strings.SelectPeripheralView.titleSyncLabel)
```
Comment (anna1901) [inline AirCasting/CreateSessionViews/SelectPeripheralView.swift:100]: Should this be the same for clearing SD card? "Select the device you'd like to sync"?

File: AirCasting/CreateSessionViews/SelectPeripheralView.swift [hunk 2]
```diff
@@ -95,7 +96,7 @@ struct SelectPeripheralView: View {
     
     var titleLabel: some View {
         var title: Text
-        if syncMode == true {
+        if syncMode == true || SDClearingRouteProcess {
```
Comment (anna1901) [inline AirCasting/CreateSessionViews/SelectPeripheralView.swift:99]: I realise it's mine, but could you please delete this "== true"?

File: AirCasting/CreateSessionViews/TurnOnBluetoothView.swift [hunk 3]
```diff
@@ -11,10 +11,12 @@ import SwiftUI
 
 struct TurnOnBluetoothView: View {
     @State private var isPowerABLinkActive = false
+    @State private var SDRestartABLink = false
     @EnvironmentObject var settingsRedirection: DefaultSettingsRedirection
     @EnvironmentObject var bluetoothManager: BluetoothManager
     @Binding var creatingSessionFlowContinues: Bool
     @Binding var sdSyncContinues: Bool
+    var isSDClearProcess: Bool
```
Comment (anna1901) [inline AirCasting/CreateSessionViews/TurnOnBluetoothView.swift:19]: Maybe this could also default to "false"? This way we wouldn't have to set it to false in all of those views above.

File: AirCasting/SDSync/ViewsFlow/SDRestartAB/SDRestartABView.swift [hunk 4]
```diff
@@ -54,15 +57,15 @@ extension SDRestartABView {
     
     var continueButton: some View {
         Button {
-            viewModel.presentNextScreen = true
+            viewModel.presentNextScreen.toggle()
```
Comment (anna1901) [inline AirCasting/SDSync/ViewsFlow/SDRestartAB/SDRestartABView.swift:60]: Why did you change "= true" to toggle?

File: AirCasting/Settings/SettingsView.swift [hunk 5]
```diff
@@ -9,29 +9,58 @@ import SwiftUI
...
     }
     
     var body: some View {
         NavigationView {
             Form {
                 signOutSection
                 settingsSection
+                #if BETA || DEBUG
+                Section() {
+                    navigateToAppConfigurationButton
```
Comment (anna1901) [inline AirCasting/Settings/SettingsView.swift:40]: Why is this in this PR? 🤔

File: AirCasting/Settings/SettingsView.swift [hunk 6]
```diff
@@ -9,29 +9,58 @@ import SwiftUI
...
             }
             .listStyle(GroupedListStyle())
             .navigationBarTitle(Strings.Settings.title)
         }
+        .fullScreenCover(isPresented: $startSDClear) {
+            CreatingSessionFlowRootView {
+                SDRestartABView(viewModel: SDRestartABViewModelDefault(urlProvider: urlProvider, isSDClearProcess: SDClearingRouteProcess), creatingSessionFlowContinues: $startSDClear)
+            }
+        }
+        .fullScreenCover(isPresented: $LocationScreenGo) {
```
Comment (anna1901) [inline AirCasting/Settings/SettingsView.swift:53]: This only works for ios15 unfortunately, so we need to check if it's ios15 or not and include that other implementation as well

File: AirCasting/SDSync/ViewsFlow/SDRestartAB/SDRestartABView.swift (hunk 4 above)
Comment (jankrzempek) [reply AirCasting/SDSync/ViewsFlow/SDRestartAB/SDRestartABView.swift:60]: It seems better/cleaner in my opinion and work as expected :)

File: AirCasting/CreateSessionViews/SelectPeripheralView.swift (hunk 1 above)
Comment (jankrzempek) [reply AirCasting/CreateSessionViews/SelectPeripheralView.swift:100]: Great catch, I will make a change then

File: AirCasting/CreateSessionViews/SelectPeripheralView.swift (hunk 2 above)
Comment (jankrzempek) [reply AirCasting/CreateSessionViews/SelectPeripheralView.swift:99]: As long as syncMode is optional, unfortunately I cannot do it

File: AirCasting/SDSync/ViewsFlow/ClearingSDCard/ClearingSDCardViewModel.swift [hunk 7]
```diff
@@ -1,18 +1,95 @@
...
+}
 
 protocol ClearingSDCardViewModel: ObservableObject {
+    var isClearingCompleted: Published<Bool>.Publisher { get }
+    var shouldDismiss: Published<Bool>.Publisher { get }
     var presentNextScreen: Bool { get set }
     var isSDClearProcess: Bool { get set }
+    func clearSDCard()
+    func getAlertTitle() -> String
+    func getAlertMessage() -> String
```
Comment (pawelgil) [inline AirCasting/SDSync/ViewsFlow/ClearingSDCard/ClearingSDCardViewModel.swift:19]: using our new system we can do `var alert: AlertInfo { get }` (if it's merged in, not sure).

If not, maybe we can omit `get` here and make it as `var`? We can use computed vars in implementation part

File: AirCasting/SDSync/ViewsFlow/ClearingSDCard/ClearingSDCardViewModel.swift [hunk 8]
```diff
@@ -1,18 +1,95 @@
...
+    case noConnection
+    case noClearing
+}
 
 protocol ClearingSDCardViewModel: ObservableObject {
+    var isClearingCompleted: Published<Bool>.Publisher { get }
+    var shouldDismiss: Published<Bool>.Publisher { get }
     var presentNextScreen: Bool { get set }
     var isSDClearProcess: Bool { get set }
+    func clearSDCard()
```
Comment (pawelgil) [inline AirCasting/SDSync/ViewsFlow/ClearingSDCard/ClearingSDCardViewModel.swift:17]: like I wrote in Janek's PR - viewmodel should not be asked to do a certain thing by the view. It should only be informed what happend view-wise (like user tapped something, switched a switch, stuff like that). So for example here it should be phrased more like `clearSDCardButtonTapped()` or smth like that.

File: AirCasting/SDSync/ViewsFlow/ClearingSDCard/ClearingSDCardView.swift [hunk 9]
```diff
@@ -68,6 +77,14 @@ extension ClearingSDCardView {
         }
     }
     
+    var alert: Alert {
```
Comment (pawelgil) [inline AirCasting/SDSync/ViewsFlow/ClearingSDCard/ClearingSDCardView.swift:80]: can use new alert system instead (hope it's merged in)

File: AirCasting/SDSync/ViewsFlow/ClearingSDCard/ClearingSDCardView.swift [hunk 10]
```diff
@@ -28,10 +30,17 @@ struct ClearingSDCardView<VM: ClearingSDCardViewModel>: View {
...
+        }
         .onAppear(perform: {
-            DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(10)) {
-                viewModel.presentNextScreen.toggle()
-            }
+            viewModel.clearSDCard()
+        })
+        .onReceive(viewModel.isClearingCompleted, perform: { result in
+            viewModel.presentNextScreen = result
+        })
```
Comment (pawelgil) [inline AirCasting/SDSync/ViewsFlow/ClearingSDCard/ClearingSDCardView.swift:41]: this `onReceive` is not needed. It is subscribed to viewmodel only to switch execution back into viewmodel. So it can be fully contained inside viewmodel implementation instead.

File: AirCasting/ABConnector/SDCardAirBeamServices.swift [hunk 11]
```diff
@@ -104,7 +105,33 @@ class BluetoothSDCardAirBeamServices: SDCardAirBeamServices {
         }
     }
     
-    func finishSync(completion: () -> Void) {
+    func clearSDCard(of peripheral: CBPeripheral, completion: @escaping (Result<Void, Error>) -> Void) {
+        metadataCharacteristicObserver = bluetoothManager.subscribeToCharacteristic(DOWNLOAD_META_DATA_FROM_SD_CARD_CHARACTERISTIC_UUID) { result in
```
Comment (pawelgil) [inline AirCasting/ABConnector/SDCardAirBeamServices.swift:109]: we are using the same `var` for `downloadData`. It can lead to problems (like if we invoke download and clear one after another the clear will overwrite the old token and it won't get unsubscribed ever). I think it's not intended?

File: AirCasting/SDSync/ViewsFlow/ClearingSDCard/ClearingSDCardViewModel.swift [hunk 12]
```diff
@@ -1,18 +1,95 @@
...
+            return Strings.ClearingSDCardView.failedClearingAlertMessage
+        case .noConnection:
+            return Strings.AirBeamConnector.connectionTimeoutDescription
+        case .noClearing:
+            return Strings.ClearingSDCardView.failedClearingAlertMessage
+        }
+    }
+    
+    private func sendClearConfig() {
+        let configurator = AirBeam3Configurator(userAuthenticationSession: self.userAuthenticationSession,
```
Comment (pawelgil) [inline AirCasting/SDSync/ViewsFlow/ClearingSDCard/ClearingSDCardViewModel.swift:91]: I'd change it so that `BluetoothSDCardAirBeamServices` sends those configurations. Both for download and for erase. We want only super high level stuff here, not some obscure configuration sending. Would that be possible?

**Review by pawelgil — CHANGES_REQUESTED** [review]

File: AirCasting/SDSync/ViewsFlow/ClearingSDCard/ClearingSDCardViewModel.swift (hunk 8 above)
Comment (jankrzempek) [reply AirCasting/SDSync/ViewsFlow/ClearingSDCard/ClearingSDCardViewModel.swift:17]: As long as it is used in the .onAppear modifier I am really not sure @pawelgil. Does `clearSDCardButtonTapped()` still work here?

File: AirCasting/ABConnector/SDCardAirBeamServices.swift (hunk 11 above)
Comment (anna1901) [reply AirCasting/ABConnector/SDCardAirBeamServices.swift:115]: Very good catch, thank you!

**Review by pawelgil — APPROVED** [review]
